File: LeetCode/2. Add Two Numbers/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* head1, ListNode* head2) {
        if (head1 == nullptr && head2 == nullptr) return nullptr;
        
        auto head = ListNode(0);
        
        auto carry = 0;
        auto current = &head;
        while (head1 != nullptr || head2 != nullptr || carry != 0) {
            auto number = carry;
            if (head1 != nullptr) number += head1->val;
            if (head2 != nullptr) number += head2->val;
            auto node = new ListNode(number);
            if (number >= 10) {
                carry = 1;
                node->val = number - 10;
            } else {
                carry = 0;
            }
            head1 = head1 != nullptr ? head1->next : nullptr;
            head2 = head2 != nullptr ? head2->next : nullptr;
            current->next = node;
            current = node;
        }

        return head.next;
    }
};

File: LeetCode/2. Add Two Numbers/Solution.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int = 0) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        val head = ListNode(0)
        
        var carry = 0
        var current = head
        var head1 = l1
        var head2 = l2
        while (head1 != null || head2 != null || carry != 0) {
            val number = (head1?.`val` ?: 0) + (head2?.`val` ?: 0) + carry
            val node = if (number >= 10) {
                carry = 1
                ListNode(number - 10)
            } else {
                carry = 0
                ListNode(number)
            }
            head1 = head1?.next
            head2 = head2?.next
            current.next = node
            current = node
        }

        
        return head.next
    }
}